wasm-runtime/polkadot/src/runtime/governance.rs

```diff
@@ -0,0 +1,361 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Governance system: Handles administration and dispatch of sensitive operations including
+//! setting new code, minting new tokens and changing parameters.
+//!
+//! For now this is limited to a simple qualified majority vote (whose parameter is retrieved from
+//! storage) between validators. A single vote may be proposed per era, and at most one approval
+//! vote may be cast by each validator. The tally is maintained through a simple tag in storage for
+//! each validator that has approved.
+//!
+//! At the end of the era, all validators approvals are tallied and if there are sufficient to pass
+//! the proposal then it is enacted. All items in storage concerning the proposal are reset.
+
+use runtime_support::prelude::*;
+use keyedvec::KeyedVec;
+use storable::{Storable, StorageVec, kill};
+use primitives::{AccountID, Hash, BlockNumber};
+use proposal::Proposal;
+use runtime::{staking, system, session};
+
+// TRANSACTION API
+
+/// Propose a sensitive action to be taken. Any action that is enactable by `Proposal` is valid.
+/// Proposal is by the `transactor` and will automatically count as an approval. Transactor must
+/// be a current validator. It is illegal to propose when there is already a proposal in effect.
+pub fn propose(validator: &AccountID, proposal: &Proposal) {
+	if Proposal::lookup(b"gov:pro").is_some() {
+		panic!("there may only be one proposal per era.");
+	}
+	proposal.store(b"gov:pro");
+	approve(validator, staking::current_era());
+}
+
+/// Approve the current era's proposal. Transactor must be a validator. This may not be done more
+/// than once for any validator in an era.
+pub fn approve(validator: &AccountID, era_index: BlockNumber) {
+	if era_index != staking::current_era() {
+		panic!("approval vote applied on non-current era.")
+	}
+	if Proposal::lookup(b"gov:pro").is_none() {
+		panic!("there must be a proposal in order to approve.");
+	}
+	if session::validators().into_iter().position(|v| &v == validator).is_none() {
+		panic!("transactor must be a validator to approve.");
+	}
+	let key = validator.to_keyed_vec(b"gov:app:");
+	if bool::lookup(&key).is_some() {
+		panic!("transactor may not approve a proposal twice in one era.");
+	}
+	true.store(&key);
+}
+
+/// Set the proportion of validators that must approve for a proposal to be enacted at the end of
+/// its era. The value, `ppm`, is measured as a fraction of 1000 rounded down to the nearest whole
+/// validator. `1000` would require the approval of all validators; `667` would require two-thirds
+/// (or there abouts) of validators.
+pub fn set_approval_ppm_required(ppm: u32) {
+	ppm.store(b"gov:apr");
```

Comment:
That looks super hacky, why there is no `context` object at least? IMHO it's super nice in rust that you see if by calling a function you're gonna modify some other object that is borrowed to it.

Reply:
well, modifying storage is a bare function. i could create a dummy context and then pointlessly hand it around from function to function, but it makes little sense to add all of that noise.